sdk/cosmos/src/errors.rs

```diff
@@ -1,45 +1,51 @@
-/// A general error having to do with Cosmos.
-#[allow(missing_docs)]
+/// An error having to do with Cosmos.
 #[non_exhaustive]
 #[derive(Debug, thiserror::Error)]
 pub enum Error {
+    /// An error as defined in the `azure_core` crate
     #[error(transparent)]
     AzureCoreError(#[from] azure_core::Error),
+    /// An error related to parsing
     #[error(transparent)]
     ParsingError(ParsingError),
-    #[error("Header not found: {0}")]
-    HeaderNotFound(String),
-    #[error("To str error: {0}")]
-    ToStrError(#[from] http::header::ToStrError),
-    #[error("http error: {0}")]
-    AzureHttpError(#[from] azure_core::HttpError),
-    #[error("stream error: {0}")]
-    StreamError(#[from] azure_core::StreamError),
+    // TODO forward this to azure_core::Error
+    #[allow(missing_docs)]
     #[error("http error: {0}")]
     HttpError(#[from] http::Error),
-    #[error("JSON error: {0}")]
-    JsonError(#[from] serde_json::Error),
-    #[error("UTF-8 conversion error: {0}")]
-    Utf8Error(#[from] std::str::Utf8Error),
-    #[error("base64 decode error: {0}")]
-    DecodeError(#[from] base64::DecodeError),
     /// Other errors that can happen but are unlikely to be matched against
     #[error(transparent)]
     Other(#[from] Box<dyn std::error::Error + Send + Sync>),
 }
 
+/// A parsing error
+///
+/// Most issues are already defined in `azure_core`
 #[derive(Debug, thiserror::Error)]
 pub enum ParsingError {
```

Comment:
Per my previous comment, does having separate errors e.g. ParsingError offer any additional value over just another enum of Error? The simpler to use the better.

Reply:
You still get that from enums, though; i.e. Error::Foo() != Error::Bar(). Hence, couldn't you have something like `Error::ParsingError()` instead of `ParsingError::Other()` and more verbose names for more specific parsing errors? I suppose implementing `From` at least makes them convertible, but it still seems like a lot of unnecessary types. In most of our SDKs, most REST errors simply roll up into a single exception/error that provides details including the HTTP status code, original message, and optionally inner errors if the service returned it. Parsing errors will most often throw/return the original exception/error.

Reply:
To be honest I'm not sure whether we should just roll open the `ParsingError` type into the main `Error` type, but I wouldn't generalize this discussion to all errors. I do think we will want Error enums to be able to refer to other error types which have more specific error information.